### PR TITLE
Tabular connectors: Add support for SalesForce and clean code

### PR DESCRIPTION
**Pull request by LucGenetier**

### CODE REVIEW

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs [hunk 1]
```diff
@@ -24,41 +24,100 @@ public sealed class SwaggerTabularService : CdpTabularService
...
+
+        // Only used for testing
+        internal static (ConnectorFunction s, ConnectorFunction c, ConnectorFunction r, ConnectorFunction u, ConnectorFunction d) GetFunctions(IReadOnlyDictionary<string, FormulaValue> globalValues, OpenApiDocument openApiDocument)
+        {
+            return new SwaggerTabularService(globalValues).GetFunctions(new PowerFxConfig(), openApiDocument);
+        }
+
+        /* Schema [GET] */
+
+        private const string MetadataServiceRegex = @"/\$metadata\.json/datasets/({[^{}]+}(,{[^{}]+})?|default)/tables/{[^{}]+}$";
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:64]: >default [](http://example.com/codeflow?start=103&length=7)

We can't assume this name.

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs [hunk 2]
```diff
@@ -24,41 +24,100 @@ public sealed class SwaggerTabularService : CdpTabularService
...
+                return new SwaggerTabularService(globalValues).LoadSwaggerAndIdentifyKeyMethods(new PowerFxConfig(), openApiDocument);
+            }
+            catch (Exception ex)
+            {
+                error = ex.Message;
+                return false;
+            }
+        }
+
+        // Only used for testing
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:56]: >// Only used for testing [](http://example.com/codeflow?start=8&length=24)

Consider an [Obsolete] to warn people off.

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs [hunk 3]
```diff
@@ -24,41 +24,100 @@ public sealed class SwaggerTabularService : CdpTabularService
         private readonly string _connectionId;
         private IReadOnlyList<ConnectorFunction> _tabularFunctions;
         private ConnectorFunction _metadataService;
+        private ConnectorFunction _createItems;
         private ConnectorFunction _getItems;
+        private ConnectorFunction _updateItem;
+        private ConnectorFunction _deleteItem;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:30]: who sets these fields?

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs (hunk 2 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:56]: Will add

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs (hunk 1 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:64]: observations tend to prove the opposite
an alternative would be to have a 2 step approach
- first try with dataset param {...} (or params)
- if no function is identified, try with default

WDYT?

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs (hunk 3 above)
Comment (LucGenetier) [reply src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:30]: internal ConnectorFunction DeleteItem => _deleteItem ??= GetFunction...

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs [hunk 4]
```diff
@@ -72,14 +131,25 @@ public async Task InitAsync(PowerFxConfig config, OpenApiDocument openApiDocumen
             }
         }
 
-        // TABLE DATA SERVICE - CREATE
-        // POST: /datasets/{datasetName}/tables/{tableName}/items?api-version=2015-09-01
+        internal bool LoadSwaggerAndIdentifyKeyMethods(PowerFxConfig config, OpenApiDocument openApiDocument, ConnectorLogger logger = null)
+        {
+            var (s, c, r, u, d) = GetFunctions(config, openApiDocument, logger);
+            return s != null && c != null && r != null && u != null && d != null;
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.Connectors/Environment/SwaggerTabularService.cs:137]: > && u != null && d != null [](http://example.com/codeflow?start=54&length=26)

could a readonly table be missing these?

**Review by MikeStall — APPROVED** [review]
:shipit:

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**